conans/client/toolchain/cmake.py

```diff
@@ -0,0 +1,305 @@
+    """)
+
+    def __init__(self, conanfile, generator=None, generator_platform=None, build_type=None,
+                 cmake_system_name=True, toolset=None, parallel=True, make_program=None,
+                 # cmake_program=None,  # TODO: cmake program should be considered in the environment
+                 ):
+        self._conanfile = conanfile
+
+        self._fpic = self._deduce_fpic()
+        self._vs_static_runtime = self._deduce_vs_static_runtime()
+
+        self._set_rpath = True
+        self._set_std = True
+        self._set_libcxx = True
+
+        # To find the generated cmake_find_package finders
+        self._cmake_prefix_path = "${CMAKE_BINARY_DIR}"
+        self._cmake_module_path = "${CMAKE_BINARY_DIR}"
+
+        self._generator = generator or get_generator(self._conanfile)
+        self._generator_platform = (generator_platform or
+                                    get_generator_platform(self._conanfile.settings,
+                                                           self._generator))
+        self._toolset = toolset or get_toolset(self._conanfile.settings, self._generator)
+        self._build_type = build_type or self._conanfile.settings.get_safe("build_type")
+
+        builder = CMakeDefinitionsBuilder(self._conanfile,
+                                          cmake_system_name=cmake_system_name,
+                                          make_program=make_program, parallel=parallel,
+                                          generator=self._generator,
+                                          set_cmake_flags=False,
+                                          output=self._conanfile.output)
+        self.definitions = Definitions()
+        self.definitions.update(builder.get_definitions())
+        # FIXME: Removing too many things. We want to bring the new logic for the toolchain here
+        # so we don't mess with the common code.
+        self.definitions.pop("CMAKE_BUILD_TYPE", None)
+        self.definitions.pop("CONAN_IN_LOCAL_CACHE", None)
+        self.definitions.pop("CMAKE_PREFIX_PATH", None)
+        self.definitions.pop("CMAKE_MODULE_PATH", None)
+        self.definitions.pop("CONAN_LINK_RUNTIME", None)
+        for install in ("PREFIX", "BINDIR", "SBINDIR", "LIBEXECDIR", "LIBDIR", "INCLUDEDIR",
+                        "OLDINCLUDEDIR", "DATAROOTDIR"):
+            self.definitions.pop("CMAKE_INSTALL_%s" % install, None)
+
+    def _deduce_fpic(self):
+        fpic = self._conanfile.options.get_safe("fPIC")
+        if fpic is None:
+            return None
+        os_ = self._conanfile.settings.get_safe("os")
+        if os_ and "Windows" in os_:
+            self._conanfile.output.error("fPIC option defined for Windows. Remove it.")
+            return None
+        shared = self._conanfile.options.get_safe("shared")
+        if shared:
+            self._conanfile.output.error("fPIC option defined for a shared library. Remove it.")
+            return None
+        return fpic
```

Comment:
Yes, agree, but lets generalize after we have another occurrence, another toolchain. I didn't want to touch other files, this is why this is completely private, will not be documented, to be able to move it later.

Reply:
We cannot hide this behavior here. There are more people contributing to Conan and we cannot expect them to realize what is happening here. OTH, recipes might be doing conditionals based on the value of this option. Options should be modified/removed after the configure/config_options, and with a warning message.

Reply:
Not only this, everything that the toolchain does right now is completely automatic and private. The cppstd, the rpaths, the libcxx. Everything is deduced from the settings/options, and the user has not interface to change that automatic behavior. This is only used to set or not set the ``CMAKE_POSITION_INDEPENDENT_CODE`` flag in the ``conan_toolchain.cmake``, it is not modifying anything, nor removing the options, just highlighting that an incorrect value reached to the toolchain. I can put a "toolchain:" in the output messages to clearly identify that is the toolchain that is complaining (in the same way I could put other messages if the toolchain doesn't get any architecture value and it needs it).

Reply:
It highlights an incorrect value (according to our understanding, but there are libraries that need this behavior), it changes the value and will build with a message. This could be opinionated, but this is not my main concern now.


My main concern is about Conan developers, trying to figure out and gather everything related to the _configure_ step. They will need to search for special logic scattered around the codebase instead of having it all together in the [client/conanfile/configure.py](https://github.com/conan-io/conan/blob/develop/conans/client/conanfile/configure.py) file:

```python
if hasattr(conanfile, "toolchain", None) or CONAN_V2_MODE:
    # ... 'fPIC' value is adjusted for the new toolchain paradigm
    ...
    # Conan v2. Should we raise ConanInvalidConfiguration? How to opt-out?
```

I really think it is a good idea to have everything related to the _configure_ step in the one single file dedicated to this step.

Reply:
I have completely changed the messages, moved them to warnings, and highlight that it is just the toolchain that will not be using this values in the toolchain.

Reply:
> I really think it is a good idea to have everything related to the configure step in the one single file dedicated to this step.

yeah, I totally agree. It wasn't my intention to put this functionality here, it was just an attempt to explain the decissions that the toolchain was taking. I think with the new warnings it will be more clear, and we need to discuss how to implement this logic at the ``configure()`` steps.

Reply:
Put the check logic in ``configure()`` time. Review it there please.